whois-nrtm4/src/main/java/net/ripe/db/nrtm4/SnapshotFileGenerator.java

```diff
@@ -0,0 +1,101 @@
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.text.DecimalFormat;
+import java.util.Optional;
+
+
+@Service
+public class SnapshotFileGenerator {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SnapshotFileGenerator.class);
+
+    private final NrtmVersionInfoRepository nrtmVersionInfoRepository;
+    private final SnapshotFileSerializer snapshotFileSerializer;
+    private final SnapshotFileRepository snapshotFileRepository;
+    private final SnapshotObjectSynchronizer snapshotObjectSynchronizer;
+    private final NrtmFileStore nrtmFileStore;
+
+    public SnapshotFileGenerator(
+        final NrtmVersionInfoRepository nrtmVersionInfoRepository,
+        final SnapshotFileSerializer snapshotFileSerializer,
+        final SnapshotFileRepository snapshotFileRepository,
+        final SnapshotObjectSynchronizer snapshotObjectSynchronizer,
+        final NrtmFileStore nrtmFileStore
+    ) {
+        this.nrtmVersionInfoRepository = nrtmVersionInfoRepository;
+        this.snapshotFileSerializer = snapshotFileSerializer;
+        this.snapshotFileRepository = snapshotFileRepository;
+        this.snapshotObjectSynchronizer = snapshotObjectSynchronizer;
+        this.nrtmFileStore = nrtmFileStore;
+    }
+
+    public Optional<PublishableSnapshotFile> createSnapshot(final NrtmSource source) {
+        final String method = "createSnapshot";
+        // Get last version from database.
+        final Optional<NrtmVersionInfo> lastVersion = nrtmVersionInfoRepository.findLastVersion(source);
+        NrtmVersionInfo version;
+        LOGGER.info("{} lastVersion.isEmpty() {}", method, lastVersion.isEmpty());
+        if (lastVersion.isEmpty()) {
+            version = snapshotObjectSynchronizer.initializeSnapshotObjects(source);
+        } else {
+            version = lastVersion.get();
+            if (version.getType() == NrtmDocumentType.DELTA) {
+                LOGGER.info("Not generating snapshot file yet");
+                return Optional.empty();
+            } else {
+                LOGGER.info("{} Not generating snapshot file since no deltas have been published since v{} with serialID {}",
+                    method, version.getVersion(), version.getLastSerialId());
+                return Optional.empty();
+            }
+        }
+        LOGGER.info("{} now at version: {}", method, version);
+        if (version.getVersion() > 1) {
+            LOGGER.debug("not syncing deltas to snapshot");
+            return Optional.empty();
+        }
+        final PublishableSnapshotFile snapshotFile = new PublishableSnapshotFile(version);
+        final String fileName = NrtmFileUtil.newFileName(snapshotFile);
+        try {
+            final long start = System.currentTimeMillis();
+            final OutputStream out = nrtmFileStore.getFileOutputStream(snapshotFile.getSessionID(), fileName);
+            snapshotFileSerializer.writeSnapshotAsJson(snapshotFile, out);
+            out.close();
```

Comment:
Use try-with-resources instead of an explicit close(), e.g. `try (final Outputstream out = ...) { ... }`